DataFormats/L1TGlobal/interface/GlobalObject.h

```diff
@@ -1,62 +1,105 @@
 #ifndef DataFormats_L1TGlobal_GlobalObject_h
 #define DataFormats_L1TGlobal_GlobalObject_h
 
-// system include files
 #include <string>
+#include <utility>
+#include <vector>
 
 namespace l1t {
 
-  // user include files
-  //   base class
+  /* Enum of L1T GlobalObjects
 
-  // forward declarations
+     IMPORTANT
+       The elements of the enum l1t::GlobalObject are used in the data format GlobalObjectMapRecord.
+       One instance of GlobalObjectMapRecord is produced online by the HLT,
+       and it is written to disk as part of the RAW data tier.
+       Said instance of GlobalObjectMapRecord is produced at HLT by the plugin L1TGlobalProducer,
+       which implements the emulator of the Stage-2 Level-1 Global Trigger.
 
-  /// L1 GT objects
-  ///    ObjNull catch all errors
+       In order not to change the meaning of existing data when adding entries to the enum l1t::GlobalObject,
+       it is necessary to add such new entries with an explicit integer value which has not been used before in the enum.
+
+       When adding new elements to the enum l1t::GlobalObject, make sure to also update accordingly
+       (a) the vector l1t::kGlobalObjectEnumStringPairs in this file, and
+       (b) the unit test implemented in test/test_catch2_l1tGlobalObject.cc in this package.
+
+       For further information on the subject, please see
+        https://github.com/cms-sw/cmssw/pull/42634#discussion_r1302636113
+        https://github.com/cms-sw/cmssw/issues/42719
+  */
   enum GlobalObject {
-    gtMu,
-    gtMuShower,
-    gtEG,
-    gtJet,
-    gtTau,
-    gtETM,
-    gtETT,
-    gtHTT,
-    gtHTM,
-    gtETMHF,
-    gtTowerCount,
-    gtMinBiasHFP0,
-    gtMinBiasHFM0,
-    gtMinBiasHFP1,
-    gtMinBiasHFM1,
-    gtETTem,
-    gtAsymmetryEt,
-    gtAsymmetryHt,
-    gtAsymmetryEtHF,
-    gtAsymmetryHtHF,
-    gtCentrality0,
-    gtCentrality1,
-    gtCentrality2,
-    gtCentrality3,
-    gtCentrality4,
-    gtCentrality5,
-    gtCentrality6,
-    gtCentrality7,
-    gtExternal,
-    gtZDCP,
-    gtZDCM,
-    ObjNull
+    gtMu = 0,
+    gtMuShower = 1,
+    gtEG = 2,
+    gtJet = 3,
+    gtTau = 4,
+    gtETM = 5,
+    gtETT = 6,
+    gtHTT = 7,
+    gtHTM = 8,
+    gtETMHF = 9,
+    gtTowerCount = 10,
+    gtMinBiasHFP0 = 11,
+    gtMinBiasHFM0 = 12,
+    gtMinBiasHFP1 = 13,
+    gtMinBiasHFM1 = 14,
+    gtETTem = 15,
+    gtAsymmetryEt = 16,
+    gtAsymmetryHt = 17,
+    gtAsymmetryEtHF = 18,
+    gtAsymmetryHtHF = 19,
+    gtCentrality0 = 20,
+    gtCentrality1 = 21,
+    gtCentrality2 = 22,
+    gtCentrality3 = 23,
+    gtCentrality4 = 24,
+    gtCentrality5 = 25,
+    gtCentrality6 = 26,
+    gtCentrality7 = 27,
+    gtExternal = 28,
+    gtZDCP = 29,
+    gtZDCM = 30,
+    ObjNull = 31,
```

Comment:
So if there is a new object added, would it be added with 32, or with 31 and ObjNull shifted to 32?

First way would have objects after ObjNull (which could pose problems in downstream codes), second would change the meaning of "31".

??

Reply:
I don't know.

Is there a guarantee that objects of type `ObjNull` will not be written to output ? Since I couldn't give this guarantee, it seemed safer to set `ObjNull` to its value (with the idea of keeping it as `31` from now on). This implies using `32` and higher for any future elements of the `enum`.

Reply:
Could we instead have the `ObjNull` be zero?

Reply:
Unfortunately, zero is taken (see code above), and changing the enum in a non-backward-compatible way is not an option (it would change the meaning of the data we took when reading it post-PR). I think that, indeed, ObjNull==0 would have been a better choice when this was designed, but that ship sailed a long time ago.